src/schema.jl

```diff
@@ -0,0 +1,84 @@
+"""
+An XML Schema Document, produced by an XML file or XMLDocument that is XML for the schema.
+"""
+mutable struct XMLSchema
+    ptr::Xptr
+    function XMLSchema(ctxt::Xptr)
+        schema = ccall((:xmlSchemaParse, libxml2), Xptr, (Xptr,), ctxt)
+        schema != C_NULL || throw(XMLValidationError("Bad XML Schema in Document"))
+        ccall((:xmlSchemaFreeParserCtxt, libxml2), Cvoid, (Xptr,), ctxt)
+        obj = new(schema)
+        finalizer((x) -> Libc.free(x.ptr), obj)
+    end
+end
+
+"""
+Create an XMLSchema from a file or url
+"""
+function XMLSchema(url::String)
+    ctxt = ccall((:xmlSchemaNewParserCtxt, libxml2), Xptr, (Cstring,), url)
+    ctxt != C_NULL || throw(XMLValidationError("Bad XML Schema at " * url))
+    return XMLSchema(ctxt)
+end
+
+"""
+Create an XMLSchema from an XMLDocument
+"""
+function XMLSchema(doc::XMLDocument)
+    ctxt = ccall((:xmlSchemaNewDocParserCtxt, libxml2), Xptr, (Xptr,), doc.ptr)
+    ctxt != C_NULL || throw(XMLValidationError("Bad XML Schema in Document"))
+    return XMLSchema(ctxt)
+end
+
+"""
+Validate an XMLDocument with an XMLSchema
+Returns true if valid
+"""
+function validate(xml::XMLDocument, schema::XMLSchema)
+    ctxt = ccall((:xmlSchemaNewValidCtxt, libxml2), Xptr, (Xptr,), schema.ptr)
+    err = ccall((:xmlSchemaValidateDoc, libxml2), 
+        Cint, (Xptr, Xptr), ctxt, xml.ptr)
+    Libc.free(ctxt)
+    return err == 0 ? true : false
+end
+
+"""
+Validate an XML file or url with an XMLSchema
+Returns true if valid
+"""
+function validate(url::String, schema::XMLSchema)
+    ctxt = ccall((:xmlSchemaNewValidCtxt, libxml2), Xptr, (Xptr,), schema.ptr)
+    err = ccall((:xmlSchemaValidateFile, libxml2), 
+        Cint, (Xptr, Cstring), ctxt, url)
+    Libc.free(ctxt)
+    return err == 0 ? true : false
+end
+
+"""
+Validate an XMLElement of an XMLDocument with an XMLSchema
+Returns true if valid
+"""
+function validate(elem::XMLElement, schema::XMLSchema)
+    ctxt = ccall((:xmlSchemaNewValidCtxt, libxml2), Xptr, (Xptr,), schema.ptr)
+    err = ccall((:xmlSchemaValidateOneElement, libxml2), 
+        Cint, (Xptr, Xptr), ctxt, elem.node.ptr)
+   Libc.free(ctxt)
+   return err == 0 ? true : false
+end
```

Comment:
There's a common pattern here where you make a new schema valid context do some work and then cleanup. We should put the common pattern into it's own function:

```julia
function _schema_valid_ctxt(f::Function, schema::XMLSchema)
    ctxt = ccall((:xmlSchemaNewValidCtxt, libxml2), Xptr, (Xptr,), schema.ptr)
    local err
    err = try
        f(ctxt)
    finally
        Libc.free(ctxt)
    end
    return err
end
```
And then it can be used via:
```julia
function validate(elem::XMLElement, schema::XMLSchema)
    err = _schema_valid_ctxt(schema) do ctxt
        ccall((:xmlSchemaValidateOneElement, libxml2), 
            Cint, (Xptr, Xptr), ctxt, elem.node.ptr)
    end
    return err == 0
end
```